src/platforms/macos/daemon/macosdnsmanager.cpp

```diff
@@ -0,0 +1,269 @@
+  Q_ASSERT(!tokens.isEmpty());
+  QString appName = tokens[0];
+
+  QCoreApplication app(CommandLineParser::argc(), CommandLineParser::argv());
+
+  QCoreApplication::setApplicationName("Mozilla VPN DNS Manager");
+  QCoreApplication::setApplicationVersion(Constants::versionString());
+  QTextStream stream(stderr);
+
+  if (tokens.length() <= 1) {
+    stream << "Error: At least one DNS server address is required." << Qt::endl;
+    stream << Qt::endl;
+    stream << "Usage: [options] ADDR [ADDR ...]" << Qt::endl;
+    return 1;
+  }
+
+  // Ensure that the DNS addresses we received are valid.
+  QStringList dnsAddrList = tokens.mid(1);
+  for (const QString& addr : dnsAddrList) {
+    QHostAddress host = QHostAddress(addr);
+    if (host.isNull()) {
+      stream << "Error: DNS Server \'" << addr << "\' is invalid." << Qt::endl;
+      return 1;
+    }
+    if (host.isBroadcast() || host.isMulticast()) {
+      stream << "Error: DNS Server \'" << addr << "\' is a multicast address."
+             << Qt::endl;
+      return 1;
+    }
+  }
+
+  // Prepare the updated DNS configuration.
+  CFMutableDictionaryRef dnsConfig = CFDictionaryCreateMutable(
+      kCFAllocatorSystemDefault, 0, &kCFCopyStringDictionaryKeyCallBacks,
+      &kCFTypeDictionaryValueCallBacks);
+  auto configGuard = qScopeGuard([&] { CFRelease(dnsConfig); });
+  cfDictSetStringList(dnsConfig, kSCPropNetDNSServerAddresses, tokens.mid(1));
+  cfDictSetString(dnsConfig, kSCPropNetDNSDomainName, "lan");
+
+  // Open the system configuration store.
+  m_scStore = SCDynamicStoreCreate(kCFAllocatorSystemDefault,
+                                   CFSTR("mozillavpn"), nullptr, nullptr);
+  if (m_scStore == nullptr) {
+    stream << "Failed to open system configuration:" << scErrorMessage()
+           << Qt::endl;
+    return 1;
+  }
+
+  // Get the list of current network services and take a snapshot.
+  for (const QString& uuid : enumerateNetServices()) {
+    stream << "Found network service: " << uuid << Qt::endl;
+    CFStringRef dnsPath = CFStringCreateWithFormat(
+        kCFAllocatorSystemDefault, nullptr,
+        CFSTR("Setup:/Network/Service/%s/DNS"), qPrintable(uuid));
+    auto pathGuard = qScopeGuard([&] { CFRelease(dnsPath); });
+
+    // Take a snapshot of the existing configuration.
+    m_snapshot[uuid] =
+        (CFDictionaryRef)SCDynamicStoreCopyValue(m_scStore, dnsPath);
+
+    // Install the updated configuration.
+    if (!SCDynamicStoreSetValue(m_scStore, dnsPath, dnsConfig)) {
+      stream << "Failed to update configuration for " << uuid << Qt::endl;
+      stream << "Error: " << scErrorMessage() << Qt::endl;
+      return 1;
+    }
+  }
+
+  // Block this process and wait for a termination signal.
+  stream << "Waiting for termination signal." << Qt::endl;
+  int sig = waitForTermination();
+  stream << "Caught signal " << sig << ". Preparing to exit." << Qt::endl;
+
+  // If we exit on SIGHUP, leave the DNS configuration in-place.
```

Comment:
Could this comment or the a comment in DnsUtilsMacos::restoreResolvers indicate that   m_dnsManager.terminate() causes SIGHUP?

Reply:
Will do, but technically this sends a `SIGTERM`

Reply:
There's nothing in the parent process that will send a `SIGHUP` I mostly just used it for testing. However, this signal in POSIX daemons is typically used to inform a daemon that its configuration has changed and may need to be updated and/or re-start itself.

I was kind of envisioning being able to use a `SIGHUP` to tell the DNS manager to change a DNS server without deactivating the VPN, but I never quite got that far.

Reply:
Would be useful to the reader if that was added as a comment. Or consider removing that code path if necessary.

Reply:
I have added a bit more to the comment, hopefully it helps.